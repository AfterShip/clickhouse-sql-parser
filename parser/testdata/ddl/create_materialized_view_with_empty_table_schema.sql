CREATE MATERIALIZED VIEW test.t0 on cluster default_cluster
ENGINE = ReplicatedAggregatingMergeTree('/clickhouse/{layer}-{shard}/test/t0', '{replica}')
PARTITION BY toYYYYMM(f0)
ORDER BY (f0)
POPULATE AS
select f0,f1,f2,coalesce(f0,f1) as f333
from
    (select
         f0,f1,f2,
         ROW_NUMBER() over(partition by f0 order by coalesce(f1,f2)) as rn
     from test.t
     where f3 in ('foo', 'bar', 'test')
       and env ='test'
    ) as tmp
where rn = 1;