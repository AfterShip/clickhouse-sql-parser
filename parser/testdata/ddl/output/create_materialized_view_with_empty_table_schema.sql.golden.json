[
  {
    "CreatePos": 0,
    "StatementEnd": 460,
    "Name": {
      "Database": {
        "Name": "test",
        "QuoteType": 1,
        "NamePos": 25,
        "NameEnd": 29
      },
      "Table": {
        "Name": "t0",
        "QuoteType": 1,
        "NamePos": 30,
        "NameEnd": 32
      }
    },
    "IfNotExists": false,
    "OnCluster": {
      "OnPos": 33,
      "Expr": {
        "Name": "default_cluster",
        "QuoteType": 1,
        "NamePos": 44,
        "NameEnd": 59
      }
    },
    "Refresh": null,
    "RandomizeFor": null,
    "DependsOn": null,
    "Settings": null,
    "HasAppend": false,
    "Engine": {
      "EnginePos": 60,
      "EngineEnd": 190,
      "Name": "ReplicatedAggregatingMergeTree",
      "Params": {
        "LeftParenPos": 99,
        "RightParenPos": 150,
        "Items": {
          "ListPos": 101,
          "ListEnd": 149,
          "HasDistinct": false,
          "Items": [
            {
              "Expr": {
                "LiteralPos": 101,
                "LiteralEnd": 136,
                "Literal": "/clickhouse/{layer}-{shard}/test/t0"
              },
              "Alias": null
            },
            {
              "Expr": {
                "LiteralPos": 140,
                "LiteralEnd": 149,
                "Literal": "{replica}"
              },
              "Alias": null
            }
          ]
        },
        "ColumnArgList": null
      },
      "PrimaryKey": null,
      "PartitionBy": {
        "PartitionPos": 152,
        "Expr": {
          "ListPos": 165,
          "ListEnd": 176,
          "HasDistinct": false,
          "Items": [
            {
              "Expr": {
                "Name": {
                  "Name": "toYYYYMM",
                  "QuoteType": 1,
                  "NamePos": 165,
                  "NameEnd": 173
                },
                "Params": {
                  "LeftParenPos": 173,
                  "RightParenPos": 176,
                  "Items": {
                    "ListPos": 174,
                    "ListEnd": 176,
                    "HasDistinct": false,
                    "Items": [
                      {
                        "Expr": {
                          "Name": "f0",
                          "QuoteType": 1,
                          "NamePos": 174,
                          "NameEnd": 176
                        },
                        "Alias": null
                      }
                    ]
                  },
                  "ColumnArgList": null
                }
              },
              "Alias": null
            }
          ]
        }
      },
      "SampleBy": null,
      "TTL": null,
      "Settings": null,
      "OrderBy": {
        "OrderPos": 178,
        "ListEnd": 190,
        "Items": [
          {
            "OrderPos": 178,
            "Expr": {
              "LeftParenPos": 187,
              "RightParenPos": 190,
              "Items": {
                "ListPos": 188,
                "ListEnd": 190,
                "HasDistinct": false,
                "Items": [
                  {
                    "Expr": {
                      "Name": "f0",
                      "QuoteType": 1,
                      "NamePos": 188,
                      "NameEnd": 190
                    },
                    "Alias": null
                  }
                ]
              },
              "ColumnArgList": null
            },
            "Alias": null,
            "Direction": ""
          }
        ]
      }
    },
    "HasEmpty": false,
    "Destination": null,
    "SubQuery": {
      "HasParen": false,
      "Select": {
        "SelectPos": 204,
        "StatementEnd": 460,
        "With": null,
        "Top": null,
        "HasDistinct": false,
        "SelectItems": [
          {
            "Expr": {
              "Name": "f0",
              "QuoteType": 1,
              "NamePos": 211,
              "NameEnd": 213
            },
            "Modifiers": [],
            "Alias": null
          },
          {
            "Expr": {
              "Name": "f1",
              "QuoteType": 1,
              "NamePos": 214,
              "NameEnd": 216
            },
            "Modifiers": [],
            "Alias": null
          },
          {
            "Expr": {
              "Name": "f2",
              "QuoteType": 1,
              "NamePos": 217,
              "NameEnd": 219
            },
            "Modifiers": [],
            "Alias": null
          },
          {
            "Expr": {
              "Name": {
                "Name": "coalesce",
                "QuoteType": 1,
                "NamePos": 220,
                "NameEnd": 228
              },
              "Params": {
                "LeftParenPos": 228,
                "RightParenPos": 234,
                "Items": {
                  "ListPos": 229,
                  "ListEnd": 234,
                  "HasDistinct": false,
                  "Items": [
                    {
                      "Expr": {
                        "Name": "f0",
                        "QuoteType": 1,
                        "NamePos": 229,
                        "NameEnd": 231
                      },
                      "Alias": null
                    },
                    {
                      "Expr": {
                        "Name": "f1",
                        "QuoteType": 1,
                        "NamePos": 232,
                        "NameEnd": 234
                      },
                      "Alias": null
                    }
                  ]
                },
                "ColumnArgList": null
              }
            },
            "Modifiers": [],
            "Alias": {
              "Name": "f333",
              "QuoteType": 1,
              "NamePos": 239,
              "NameEnd": 243
            }
          }
        ],
        "From": {
          "FromPos": 244,
          "Expr": {
            "Table": {
              "TablePos": 253,
              "TableEnd": 447,
              "Alias": null,
              "Expr": {
                "Expr": {
                  "HasParen": true,
                  "Select": {
                    "SelectPos": 254,
                    "StatementEnd": 433,
                    "With": null,
                    "Top": null,
                    "HasDistinct": false,
                    "SelectItems": [
                      {
                        "Expr": {
                          "Name": "f0",
                          "QuoteType": 1,
                          "NamePos": 270,
                          "NameEnd": 272
                        },
                        "Modifiers": [],
                        "Alias": null
                      },
                      {
                        "Expr": {
                          "Name": "f1",
                          "QuoteType": 1,
                          "NamePos": 273,
                          "NameEnd": 275
                        },
                        "Modifiers": [],
                        "Alias": null
                      },
                      {
                        "Expr": {
                          "Name": "f2",
                          "QuoteType": 1,
                          "NamePos": 276,
                          "NameEnd": 278
                        },
                        "Modifiers": [],
                        "Alias": null
                      },
                      {
                        "Expr": {
                          "Function": {
                            "Name": {
                              "Name": "ROW_NUMBER",
                              "QuoteType": 1,
                              "NamePos": 289,
                              "NameEnd": 299
                            },
                            "Params": {
                              "LeftParenPos": 299,
                              "RightParenPos": 300,
                              "Items": {
                                "ListPos": 300,
                                "ListEnd": 300,
                                "HasDistinct": false,
                                "Items": []
                              },
                              "ColumnArgList": null
                            }
                          },
                          "OverPos": 302,
                          "OverExpr": {
                            "LeftParenPos": 306,
                            "RightParenPos": 347,
                            "PartitionBy": {
                              "PartitionPos": 306,
                              "Expr": {
                                "ListPos": 320,
                                "ListEnd": 322,
                                "HasDistinct": false,
                                "Items": [
                                  {
                                    "Expr": {
                                      "Name": "f0",
                                      "QuoteType": 1,
                                      "NamePos": 320,
                                      "NameEnd": 322
                                    },
                                    "Alias": null
                                  }
                                ]
                              }
                            },
                            "OrderBy": {
                              "OrderPos": 323,
                              "ListEnd": 346,
                              "Items": [
                                {
                                  "OrderPos": 323,
                                  "Expr": {
                                    "Name": {
                                      "Name": "coalesce",
                                      "QuoteType": 1,
                                      "NamePos": 332,
                                      "NameEnd": 340
                                    },
                                    "Params": {
                                      "LeftParenPos": 340,
                                      "RightParenPos": 346,
                                      "Items": {
                                        "ListPos": 341,
                                        "ListEnd": 346,
                                        "HasDistinct": false,
                                        "Items": [
                                          {
                                            "Expr": {
                                              "Name": "f1",
                                              "QuoteType": 1,
                                              "NamePos": 341,
                                              "NameEnd": 343
                                            },
                                            "Alias": null
                                          },
                                          {
                                            "Expr": {
                                              "Name": "f2",
                                              "QuoteType": 1,
                                              "NamePos": 344,
                                              "NameEnd": 346
                                            },
                                            "Alias": null
                                          }
                                        ]
                                      },
                                      "ColumnArgList": null
                                    }
                                  },
                                  "Alias": null,
                                  "Direction": ""
                                }
                              ]
                            },
                            "Frame": null
                          }
                        },
                        "Modifiers": [],
                        "Alias": {
                          "Name": "rn",
                          "QuoteType": 1,
                          "NamePos": 352,
                          "NameEnd": 354
                        }
                      }
                    ],
                    "From": {
                      "FromPos": 360,
                      "Expr": {
                        "Table": {
                          "TablePos": 365,
                          "TableEnd": 371,
                          "Alias": null,
                          "Expr": {
                            "Database": {
                              "Name": "test",
                              "QuoteType": 1,
                              "NamePos": 365,
                              "NameEnd": 369
                            },
                            "Table": {
                              "Name": "t",
                              "QuoteType": 1,
                              "NamePos": 370,
                              "NameEnd": 371
                            }
                          },
                          "HasFinal": false
                        },
                        "StatementEnd": 371,
                        "SampleRatio": null,
                        "HasFinal": false
                      }
                    },
                    "ArrayJoin": null,
                    "Window": null,
                    "Prewhere": null,
                    "Where": {
                      "WherePos": 377,
                      "Expr": {
                        "LeftExpr": {
                          "LeftExpr": {
                            "Name": "f3",
                            "QuoteType": 1,
                            "NamePos": 383,
                            "NameEnd": 385
                          },
                          "Operation": "IN",
                          "RightExpr": {
                            "LeftParenPos": 389,
                            "RightParenPos": 410,
                            "Items": {
                              "ListPos": 391,
                              "ListEnd": 409,
                              "HasDistinct": false,
                              "Items": [
                                {
                                  "Expr": {
                                    "LiteralPos": 391,
                                    "LiteralEnd": 394,
                                    "Literal": "foo"
                                  },
                                  "Alias": null
                                },
                                {
                                  "Expr": {
                                    "LiteralPos": 398,
                                    "LiteralEnd": 401,
                                    "Literal": "bar"
                                  },
                                  "Alias": null
                                },
                                {
                                  "Expr": {
                                    "LiteralPos": 405,
                                    "LiteralEnd": 409,
                                    "Literal": "test"
                                  },
                                  "Alias": null
                                }
                              ]
                            },
                            "ColumnArgList": null
                          },
                          "HasGlobal": false,
                          "HasNot": false
                        },
                        "Operation": "AND",
                        "RightExpr": {
                          "LeftExpr": {
                            "Name": "env",
                            "QuoteType": 1,
                            "NamePos": 423,
                            "NameEnd": 426
                          },
                          "Operation": "=",
                          "RightExpr": {
                            "LiteralPos": 429,
                            "LiteralEnd": 433,
                            "Literal": "test"
                          },
                          "HasGlobal": false,
                          "HasNot": false
                        },
                        "HasGlobal": false,
                        "HasNot": false
                      }
                    },
                    "GroupBy": null,
                    "WithTotal": false,
                    "Having": null,
                    "OrderBy": null,
                    "LimitBy": null,
                    "Limit": null,
                    "Settings": null,
                    "Format": null,
                    "UnionAll": null,
                    "UnionDistinct": null,
                    "Except": null
                  }
                },
                "AliasPos": 444,
                "Alias": {
                  "Name": "tmp",
                  "QuoteType": 1,
                  "NamePos": 444,
                  "NameEnd": 447
                }
              },
              "HasFinal": false
            },
            "StatementEnd": 447,
            "SampleRatio": null,
            "HasFinal": false
          }
        },
        "ArrayJoin": null,
        "Window": null,
        "Prewhere": null,
        "Where": {
          "WherePos": 448,
          "Expr": {
            "LeftExpr": {
              "Name": "rn",
              "QuoteType": 1,
              "NamePos": 454,
              "NameEnd": 456
            },
            "Operation": "=",
            "RightExpr": {
              "NumPos": 459,
              "NumEnd": 460,
              "Literal": "1",
              "Base": 10
            },
            "HasGlobal": false,
            "HasNot": false
          }
        },
        "GroupBy": null,
        "WithTotal": false,
        "Having": null,
        "OrderBy": null,
        "LimitBy": null,
        "Limit": null,
        "Settings": null,
        "Format": null,
        "UnionAll": null,
        "UnionDistinct": null,
        "Except": null
      }
    },
    "Populate": true,
    "Comment": null,
    "Definer": null,
    "SQLSecurity": ""
  }
]